compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -467,7 +467,18 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
       case tp =>
         val parts = children.map { sym =>
           val sym1 = if (sym.is(ModuleClass)) sym.sourceModule else sym
-          ctx.typer.refineUsingParent(tp, sym1)
+          val refined = ctx.typer.refineUsingParent(tp, sym1)
+          val inhabited = new TypeAccumulator[Boolean] {
+            override def apply(x: Boolean, tp: Type) = x && {
+              tp match {
+                case AndType(tp1, tp2) => ctx.typeComparer.intersecting(tp1, tp2)
+                case OrType(tp1, tp2) => foldOver(x, tp1) || foldOver(x, tp2)
+                case _ => foldOver(x, tp)
+              }
+            }
+          }
+          if (inhabited.apply(true, refined)) refined
+          else NoType
```

Comment:
It seems that a nested non-inhabitable intersection type does not necessarily mean the outer type is non-inhabited, just as `Nothing` does not imply `F[Nothing]` is non-inhabited..

Reply:
Good point, I've updated that check to be more conservative while traversing `refined`.